libafl_qemu/src/drcov.rs

```diff
@@ -0,0 +1,145 @@
+where
+    S: UsesInput + HasMetadata,
+{
+    fn init_hooks<QT>(&self, hooks: &QemuHooks<'_, QT, S>)
+    where
+        QT: QemuHelperTuple<S>,
+    {
+        hooks.blocks_raw(Some(gen_unique_block_ids::<QT, S>), Some(trace_block));
+    }
+
+    fn pre_exec(&mut self, _emulator: &Emulator, _input: &S::Input) {}
+
+    fn post_exec(&mut self, emulator: &Emulator, _input: &S::Input) {
+        if DRCOV_IDS.lock().unwrap().len() > self.drcov_len {
+            println!("New DrCov lenght = {}", DRCOV_IDS.lock().unwrap().len());
+            let mut drcov_vec = Vec::<DrCovBasicBlock>::new();
+            for id in DRCOV_IDS.lock().unwrap().iter() {
+                for (pc, idm) in DRCOV_MAP.lock().unwrap().iter() {
+                    if *idm == *id {
+                        let block = pc2basicblock(*pc, emulator);
+                        let mut block_len = 0;
+                        for instr in &block {
+                            block_len += instr.insn_len;
+                        }
+                        block_len -= &block.last().unwrap().insn_len;
+                        drcov_vec
+                            .push(DrCovBasicBlock::new(*pc as usize, *pc as usize + block_len));
+                    }
+                }
+            }
+
+            let mut rangemap = RangeMap::<usize, (u16, String)>::new();
+            rangemap.insert(
+                (GuestAddr::MIN as usize)..(GuestAddr::MAX as usize),
+                (0, "test".to_string()),
+            );
+            DrCovWriter::new(&rangemap)
+                .write(&self.filename, &drcov_vec)
+                .expect("Failed to write coverage file");
+        }
+        self.drcov_len = DRCOV_IDS.lock().unwrap().len();
+    }
+}
+
+pub fn gen_unique_block_ids<QT, S>(
+    hooks: &mut QemuHooks<'_, QT, S>,
+    state: Option<&mut S>,
+    pc: GuestAddr,
+) -> Option<u64>
+where
+    S: HasMetadata,
+    S: UsesInput,
+    QT: QemuHelperTuple<S>,
+{
+    if let Some(h) = hooks.helpers().match_first_type::<QemuDrCovHelper>() {
+        if !h.must_instrument(pc.into()) {
+            return None;
+        }
+    }
+    let state = state.expect("The gen_unique_block_ids hook works only for in-process fuzzing");
+    if state.metadata().get::<QemuDrCovMetadata>().is_none() {
+        state.add_metadata(QemuDrCovMetadata::new());
+    }
+    let meta = state.metadata_mut().get_mut::<QemuDrCovMetadata>().unwrap();
+
+    match DRCOV_MAP.lock().unwrap().entry(pc) {
+        Entry::Occupied(e) => {
+            let id = *e.get();
+            Some(id)
+        }
+        Entry::Vacant(e) => {
+            let id = meta.current_id;
+            e.insert(id);
+            meta.current_id = id + 1;
+            // GuestAddress is u32 for 32 bit guests
+            #[allow(clippy::unnecessary_cast)]
+            Some(id as u64)
+        }
+    }
+}
+
+pub extern "C" fn trace_block(id: u64, _data: u64) {
+    DRCOV_IDS.lock().unwrap().insert(id);
```

Comment:
This one should use a vec instead of a set, that way you can see a heatmap of how often blocks have been covered

Reply:
Why would I want to know how often a block has been covered? I want to create a DrCov which only cares about if a block was hit, right?

Reply:
If that's what you want? Usually a drcov trace is an actual trace of all blocks, and it gives you additional insights like said heat map, or even just the actual path through the program.
Counter question, why would you _not_ want the full trace?

Reply:
It adds a lot of overhead in my opinion. As you and @andreafioraldi mention there is no need to have execution hooks at all. Therefore counting the executions would add a none-neglectable overhead, no?

Reply:
Imho the overhead for drcov isn't relevant, you won't do this during Fuzzing but as post processing stage (I think?)

Reply:
Would you? I mean if you only need the generation hooks you could run it during the normal fuzzing campaign with almost no overhead and get live DrCov feedback. But I guess writing the corpus to memory and tracing DrCov afterwards is also possible. Maybe I will just make this an option for DrCov!?

Reply:
Even during Fuzzing you only want to do it for new intersting testcases, so comparably low overhead

Reply:
Could be cool to support a fast and a slow mode? Since you already wrote the slow version, anyway

Reply:
I will give it a shot